elastic/elastic-package
# Enroll cmd

### Summary

Add a command that spawns a agent with preconfigured policy. The mechanism extends the system tests approach to have a configuration that accepts definition of multiple data streams. Agents are created in unique container under the `elastic-package-service` project and can be run in parallel.

**Usage `elastic-package enroll --config default_config.yml`**

The goal was a working POC for spacetime week so I didn't test this extensively, only a couple scenarios in the stack packages. I also went for quick wins in implementations by tweaking existing mechanisms like the custom agent to allow our new use case.

The initial use case is to allow automated installation of a package (see https://github.com/elastic/elastic-package/issues/952) but the mechanism could be reused for further testing scenarios, like spinning a specific version of an agent (maybe allow local binaries as well?) or replace the current system tests approach. For the latter the thinking is that the Enroll mechanism could allow parallel test execution for packages that have large sets of streams + variants (see https://github.com/elastic/integrations/pull/4442#pullrequestreview-1146776038) and speed things up, but there is maybe a simpler way for these gains.

### Process
- Agent policy is created from the configuration[1] provided to the command
- Agent is then spawned with the previously policy attached so it starts to collect metrics as soon as it is setup
- todo: cleanup agent, policy and streams when done

### Testing
- create a `default_config.yml` under `<integrations-repo>/packages/elasticsearch/_dev/policy` with the following configuration

_[1] example configuration_
```
policy_templates:
  - name: elasticsearch
    inputs:
      - type: elasticsearch/metrics
        vars:
          hosts:
            - https://elasticsearch:9200
          username: elastic
          password: changeme
        streams:
          - dataset: elasticsearch.stack_monitoring.cluster_stats
          - dataset: elasticsearch.stack_monitoring.index_recovery
          - dataset: elasticsearch.stack_monitoring.node_stats
```
- start a stack `elastic-package stack up -v -d --version 8.7.0-SNAPSHOT`
- within the `elasticsearch` package scope, enroll an agent with the configuration `elastic-package enroll --config default_config.yml`
- an agent is spawned under `elastic-package-service` project
- navigate to Fleet UI and verify a corresponding integration was created. It should be named `elastic-package_elasticsearch_<timestamp>`
